File: frontend/demo/component/popover/popover-notification-panel.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/button';
import '@vaadin/horizontal-layout';
import '@vaadin/icon';
import '@vaadin/message-list';
import '@vaadin/popover';
import '@vaadin/tabsheet';
import '@vaadin/vaadin-lumo-styles/vaadin-iconset.js';
import { format, subMinutes } from 'date-fns';
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import type { MessageListItem } from '@vaadin/message-list';
import { getPeople } from 'Frontend/demo/domain/DataService';
import { applyTheme } from 'Frontend/demo/theme';

@customElement('popover-notifications-panel')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  @state()
  allNotifications: MessageListItem[] = [];

  @state()
  unreadNotifications: MessageListItem[] = [];

  protected override async firstUpdated() {
    const { people } = await getPeople({ count: 4 });

    this.unreadNotifications = [
      {
        userName: `${people[0].firstName} ${people[0].lastName}`,
        userImg: people[0].pictureUrl,
        time: format(subMinutes(new Date(), 20), 'HH:mm'),
        text: 'Could you send me the latest TPS report from the ACME project?',
      },
      {
        userName: `${people[1].firstName} ${people[1].lastName}`,
        userImg: people[1].pictureUrl,
        time: format(subMinutes(new Date(), 30), 'HH:mm'),
        text: 'Hey, are we on track for the trade show next month?',
      },
      {
        userName: `${people[2].firstName} ${people[2].lastName}`,
        userImg: people[2].pictureUrl,
        time: format(subMinutes(new Date(), 35), 'HH:mm'),
        text: `TPS reports look good! I'm going to pass it on to Alliyah next.`,
      },
    ];
    this.allNotifications = [
      ...this.unreadNotifications,
      {
        userName: `${people[3].firstName} ${people[3].lastName}`,
        userImg: people[3].pictureUrl,
        time: format(subMinutes(new Date(), 55), 'HH:mm'),
        text: 'Hi, are you going to attend the brainstorming session tomorrow?',
      },
    ];
  }

  protected override render() {
    return html`
      <vaadin-button id="show-notifications" aria-label="Notifications" theme="icon">
        <vaadin-icon icon="lumo:bell"></vaadin-icon>
      </vaadin-button>
      <!-- tag::snippet[] -->
      <vaadin-popover
        for="show-notifications"
        theme="arrow no-padding"
        modal
        aria-labelledby="notifications-heading"
        width="300px"
        position="bottom"
      >
        ${this.renderNotifications(this.unreadNotifications, this.allNotifications)}
      </vaadin-popover>
      <!-- end::snippet[] -->
    `;
  }

  renderNotifications(unread: MessageListItem[], all: MessageListItem[]) {
    return html`
      <vaadin-horizontal-layout
        style="align-items: center; padding: var(--vaadin-padding-l) var(--vaadin-padding-l) var(--vaadin-padding-s)"
      >
        <h4 style="margin: 0" id="notifications-heading">Notifications</h4>
        <vaadin-button theme="small" style="margin: 0 0 0 auto;" @click="${this.markAllRead}">
          Mark all read
        </vaadin-button>
      </vaadin-horizontal-layout>
      <vaadin-tabsheet class="notifications" theme="small no-padding no-border">
        <vaadin-tabs slot="tabs">
          <vaadin-tab id="unread-tab">Unread</vaadin-tab>
          <vaadin-tab id="all-tab">All</vaadin-tab>
        </vaadin-tabs>
        <div tab="unread-tab">
          ${unread.length
            ? html`<vaadin-message-list .items="${unread}"></vaadin-message-list>`
            : html`<div class="no-notifications-msg">No unread notifications</div>`}
        </div>
        <div tab="all-tab">
          <vaadin-message-list .items="${all}"></vaadin-message-list>
        </div>
      </vaadin-tabsheet>
    `;
  }

  markAllRead() {
    this.unreadNotifications = [];
  }
}

File: src/main/java/com/vaadin/demo/component/popover/PopoverNotificationPanel.java
package com.vaadin.demo.component.popover;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.Arrays;
import java.util.List;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.DataService;
import com.vaadin.demo.domain.Person;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.button.ButtonVariant;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.H4;
import com.vaadin.flow.component.messages.MessageList;
import com.vaadin.flow.component.messages.MessageListItem;
import com.vaadin.flow.component.orderedlayout.FlexComponent;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.popover.Popover;
import com.vaadin.flow.component.popover.PopoverPosition;
import com.vaadin.flow.component.popover.PopoverVariant;
import com.vaadin.flow.component.tabs.TabSheet;
import com.vaadin.flow.component.tabs.TabSheetVariant;
import com.vaadin.flow.theme.lumo.LumoIcon;
import com.vaadin.flow.router.Route;

@Route("popover-notification-panel")
public class PopoverNotificationPanel extends Div {

    public PopoverNotificationPanel() {
        Button button = new Button(LumoIcon.BELL.create());
        button.addThemeVariants(ButtonVariant.LUMO_ICON);
        button.setAriaLabel("Notifications");

        // tag::snippet[]
        Popover popover = new Popover();
        popover.setTarget(button);
        popover.setWidth("300px");
        popover.addThemeVariants(PopoverVariant.ARROW,
                PopoverVariant.LUMO_NO_PADDING);
        popover.setPosition(PopoverPosition.BOTTOM);
        popover.setModal(true);
        popover.setAriaLabelledBy("notifications-heading");
        // end::snippet[]

        List<Person> people = DataService.getPeople(5);

        MessageList unreadList = new MessageList();

        MessageList allList = new MessageList();

        MessageListItem message1 = new MessageListItem(
                "Could you send me the latest TPS report from the ACME project?",
                minutesAgo(20), people.get(0).getFullName(),
                people.get(0).getPictureUrl());

        MessageListItem message2 = new MessageListItem(
                "Hey, are we on track for the trade show next month?",
                minutesAgo(30), people.get(1).getFullName(),
                people.get(1).getPictureUrl());

        MessageListItem message3 = new MessageListItem(
                "TPS reports look good! I'm going to pass it on to Alliyah next.",
                minutesAgo(35), people.get(2).getFullName(),
                people.get(2).getPictureUrl());

        MessageListItem message4 = new MessageListItem(
                "Hi, are you going to attend the brainstorming session tomorrow?",
                minutesAgo(55), people.get(3).getFullName(),
                people.get(3).getPictureUrl());

        unreadList.setItems(Arrays.asList(message1, message2, message3));
        allList.setItems(Arrays.asList(message1, message2, message3, message4));

        TabSheet tabSheet = new TabSheet();
        tabSheet.addThemeVariants(TabSheetVariant.LUMO_TABS_SMALL,
                TabSheetVariant.LUMO_NO_PADDING);
        tabSheet.addClassName("notifications");

        Div unreadContent = new Div();
        unreadContent.add(unreadList);

        tabSheet.add("Unread", unreadContent);
        tabSheet.add("All", new Div(allList));

        H4 heading = new H4("Notifications");
        heading.setId("notifications-heading");
        heading.getStyle().set("margin", "0");

        Button markRead = new Button("Mark all read", (e) -> {
            unreadContent.removeAll();
            unreadContent.add(new Div("No new notifications") {
                {
                    this.addClassName("no-notifications-msg");
                }
            });
        });
        markRead.getStyle().set("margin", "0 0 0 auto");
        markRead.addThemeVariants(ButtonVariant.LUMO_SMALL);

        HorizontalLayout layout = new HorizontalLayout(heading, markRead);
        layout.setSpacing(false);
        layout.setAlignItems(FlexComponent.Alignment.CENTER);
        layout.getStyle().set("padding",
                "1rem 1rem 0.25rem");

        popover.add(layout, tabSheet);

        add(button, popover);
    }

    private Instant minutesAgo(int minutes) {
        return LocalDateTime.now(ZoneOffset.UTC).minusMinutes(minutes)
                .toInstant(ZoneOffset.UTC);
    }

    public static class Exporter
            extends DemoExporter<PopoverNotificationPanel> { // hidden-source-line
    } // hidden-source-line
}
